// @copilotkit/react-core/dist/chunk-4OIVQMEM.mjs
import {
  defaultCopilotContextCategories
} from "./chunk-LLU5URI5.mjs";
import {
  fetchAndDecodeChatCompletion
} from "./chunk-MJKBCG4U.mjs";
import {
  __async,
  __spreadValues
} from "./chunk-SKC7AJIV.mjs";

// src/utils/extract.ts
import {
  COPILOT_CLOUD_PUBLIC_API_KEY_HEADER
} from "@copilotkit/shared";
import untruncateJson from "untruncate-json";
function extract(_0) {
  return __async(this, arguments, function* ({
    context,
    instructions,
    parameters,
    include,
    data,
    abortSignal,
    stream
  }) {
    var _a, _b;
    const { messages } = context;
    const action = {
      name: "extract",
      parameters,
      handler: (args) => {
      }
    };
    const includeReadable = (_a = include == null ? void 0 : include.readable) != null ? _a : false;
    const includeMessages = (_b = include == null ? void 0 : include.messages) != null ? _b : false;
    let contextString = "";
    if (data) {
      contextString = (typeof data === "string" ? data : JSON.stringify(data)) + "\n\n";
    }
    if (includeReadable) {
      contextString += context.getContextString([], defaultCopilotContextCategories);
    }
    const systemMessage = {
      id: "system",
      content: makeSystemMessage(contextString, instructions),
      role: "system"
    };
    const headers = __spreadValues(__spreadValues({}, context.copilotApiConfig.headers || {}), context.copilotApiConfig.publicApiKey ? { [COPILOT_CLOUD_PUBLIC_API_KEY_HEADER]: context.copilotApiConfig.publicApiKey } : {});
    const response = yield fetchAndDecodeChatCompletion({
      copilotConfig: context.copilotApiConfig,
      messages: includeMessages ? [systemMessage, ...messages] : [systemMessage],
      tools: context.getChatCompletionFunctionDescriptions({ extract: action }),
      headers,
      body: context.copilotApiConfig.body,
      toolChoice: { type: "function", function: { name: "extract" } },
      signal: abortSignal
    });
    if (!response.events) {
      throw new Error("extract() failed: Could not fetch chat completion");
    }
    const reader = response.events.getReader();
    let isInitial = true;
    while (true) {
      const { done, value } = yield reader.read();
      if (done) {
        break;
      }
      if (value.type === "partial") {
        try {
          let partialArguments = JSON.parse(untruncateJson(value.arguments));
          stream == null ? void 0 : stream({
            status: isInitial ? "initial" : "inProgress",
            args: partialArguments
          });
          isInitial = false;
        } catch (e) {
        }
      }
      if (value.type === "function") {
        stream == null ? void 0 : stream({
          status: "complete",
          args: value.arguments
        });
        return value.arguments;
      }
    }
    throw new Error("extract() failed: No function call occurred");
  });
}
function makeSystemMessage(contextString, instructions) {
  return `
Please act as an efficient, competent, conscientious, and industrious professional assistant.

Help the user achieve their goals, and you do so in a way that is as efficient as possible, without unnecessary fluff, but also without sacrificing professionalism.
Always be polite and respectful, and prefer brevity over verbosity.

The user has provided you with the following context:
\`\`\`
${contextString}
\`\`\`

They have also provided you with a function you MUST call to initiate actions on their behalf.

Please assist them as best you can.

This is not a conversation, so please do not ask questions. Just call the function without saying anything else.

The user has given you the following task to complete:

\`\`\`
${instructions}
\`\`\`

Any additional messages provided are for providing context only and should not be used to ask questions or engage in conversation.
`;
}

export {
  extract
};
