// @copilotkit/react-ui/dist/chunk-BJPGMY3I.mjs
import {
  ActivityIcon,
  CloseIcon,
  HeaderCloseIcon,
  OpenIcon,
  PushToTalkIcon,
  RegenerateIcon,
  SendIcon,
  SpinnerIcon,
  StopIcon
} from "./chunk-FZC7X5PK.mjs";
import {
  __spreadValues
} from "./chunk-MRXNTQOX.mjs";

// src/components/chat/ChatContext.tsx
import React from "react";
import { jsx } from "react/jsx-runtime";
var ChatContext = React.createContext(void 0);
function useChatContext() {
  const context = React.useContext(ChatContext);
  if (context === void 0) {
    throw new Error(
      "Context not found. Did you forget to wrap your app in a <ChatContextProvider> component?"
    );
  }
  return context;
}
var ChatContextProvider = ({
  // temperature,
  // instructions,
  // maxFeedback,
  labels,
  icons,
  children,
  open,
  setOpen
}) => {
  const context = {
    labels: __spreadValues(__spreadValues({}, {
      initial: "",
      title: "CopilotKit",
      placeholder: "Type a message...",
      error: "\u274C An error occurred. Please try again.",
      stopGenerating: "Stop generating",
      regenerateResponse: "Regenerate response"
    }), labels),
    icons: __spreadValues(__spreadValues({}, {
      openIcon: OpenIcon,
      closeIcon: CloseIcon,
      headerCloseIcon: HeaderCloseIcon,
      sendIcon: SendIcon,
      activityIcon: ActivityIcon,
      spinnerIcon: SpinnerIcon,
      stopIcon: StopIcon,
      regenerateIcon: RegenerateIcon,
      pushToTalkIcon: PushToTalkIcon
    }), icons),
    open,
    setOpen
  };
  return /* @__PURE__ */ jsx(ChatContext.Provider, { value: context, children });
};

export {
  ChatContext,
  useChatContext,
  ChatContextProvider
};
